pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,267 @@
+	}
+
+	var pgInfo1 PodGroupInfo
+	var pgInfo2 PodGroupInfo
+	var ok bool
+	podGroupName1, min1, namespace1 := cs.checkPod(pod1)
+	podGroupName2, min2, namespace2 := cs.checkPod(pod2)
+	if podGroupName1 != "" {
+		key1 := namespace1 + "/" + podGroupName1
+		pgInfo1, ok = cs.PodGroupInfos[key1]
+		if !ok {
+			pgInfo1 = PodGroupInfo{
+				InitialTimestamp: pod1.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min1,
+			}
+			cs.PodGroupInfos[key1] = pgInfo1
+		}
+	}
+	if podGroupName2 != "" {
+		key2 := namespace2 + "/" + podGroupName2
+		pgInfo2, ok = cs.PodGroupInfos[key2]
+		if !ok {
+			pgInfo2 = PodGroupInfo{
+				InitialTimestamp: pod2.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min2,
+			}
+			cs.PodGroupInfos[key2] = pgInfo2
+		}
+	}
+
+	if podGroupName1 != "" && podGroupName2 != "" {
+		if pgInfo1.InitialTimestamp.Equal(pgInfo2.InitialTimestamp) {
+			return pgInfo1.Id < pgInfo2.Id
+		} else {
+			return pgInfo1.InitialTimestamp.Before(pgInfo2.InitialTimestamp)
+		}
+	} else if podGroupName1 != "" {
+		return pgInfo1.InitialTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+	} else if podGroupName2 != "" {
+		return podInfo1.InitialAttemptTimestamp.Before(pgInfo2.InitialTimestamp)
+	}
+
+	return podInfo1.InitialAttemptTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+}
+
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable, namespace := cs.checkPod(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, "")
+	}
+
+	total := cs.calculateTotalPods(podGroupName, namespace)
+	if total < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in PreFilter: %d",
+			namespace+"/"+podGroupName, minAvailable, total)
+		return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+func (cs *Coscheduling) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Permit is the functions invoked by the framework at "permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	podGroupName, minAvailable, namespace := cs.checkPod(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, ""), 0
+	}
+
+	running := cs.calculateRunningPods(podGroupName, namespace)
+	waiting := cs.calculateWaitingPods(podGroupName, namespace)
+
+	if running+waiting+1 < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+			namespace+"/"+podGroupName, minAvailable, running, waiting)
+		return framework.NewStatus(framework.Wait, ""), time.Duration(minAvailable-running-waiting-1/minAvailable) * PermitWaitingTime
```

Comment:
what's the point of `1/minAvailable`? I suppose it's evaluated to 0 when `minAvailable` >= 1 - which I think is the common case.